imports/server/JoinPublisher.ts

```diff
@@ -0,0 +1,266 @@
+    this.sub = sub;
+    this.spec = spec;
+    this.observers = observers;
+  }
+
+  incref(id: string) {
+    if (this.subscribers.has(id)) {
+      this.subscribers.get(id)!.refCount += 1;
+    } else {
+      this.subscribers.set(id, {
+        refCount: 1,
+        subscriber: new JoinedObjectObserver(this.sub, this.spec, id, this.observers),
+      });
+    }
+  }
+
+  decref(id: string) {
+    const record = this.subscribers.get(id);
+    if (!record) {
+      return;
+    }
+    record.refCount -= 1;
+    if (record.refCount <= 0) {
+      record.subscriber.destroy();
+      this.subscribers.delete(id);
+    }
+  }
+
+  shutdown() {
+    this.subscribers.forEach((v) => v.subscriber.destroy());
+  }
+}
+
+class JoinedObjectObserver<T extends { _id: string }> {
+  sub: Subscription;
+
+  id: string;
+
+  modelName: string;
+
+  foreignKeys?: PublishSpec<T>['foreignKeys'];
+
+  watcher: Meteor.LiveQueryHandle;
+
+  exists: boolean = false;
+
+  observers: Map<string, RefCountedJoinedObjectObserverMap<any>>;
+
+  values: Map<string, Map<string, string>> = new Map();
+
+  constructor(
+    sub: Subscription,
+    spec: PublishSpec<T>,
+    id: string,
+    observers: Map<string, RefCountedJoinedObjectObserverMap<any>>
+  ) {
+    const { model, projection, foreignKeys } = spec;
+
+    this.sub = sub;
+    this.id = id;
+    this.modelName = model._name;
+    this.foreignKeys = foreignKeys;
+    this.observers = observers;
+
+    this.watcher = model.find(id, { fields: projection as Mongo.FieldSpecifier }).observeChanges({
+      added: (_, fields) => {
+        const fkValues = new Map<string, string>();
+        foreignKeys?.forEach(({ field, join }) => {
+          const val = fields[field] as unknown as string;
+          if (!val) {
+            return;
+          }
+
+          this.observers.get(join.model._name)!.incref(val);
+          fkValues.set(field, val);
+        });
+
+        this.sub.added(this.modelName, id, fields);
+        this.exists = true;
+        this.values.set(id, fkValues);
+      },
+      changed: (_, fields) => {
+        // add new foreign keys first
+        foreignKeys?.forEach(({ field, join }) => {
+          const val = fields[field] as unknown as string;
+          if (!val) {
+            return;
+          }
+
+          this.observers.get(join.model._name)!.incref(val);
+        });
+        this.sub.changed(this.modelName, id, fields);
+
+        // then remove old foreign key values
+        const fkValues = this.values.get(id)!;
+        foreignKeys?.forEach(({ field, join }) => {
+          const val = fkValues.get(field);
+          if (!val) {
+            return;
+          }
+
+          this.observers.get(join.model._name)!.decref(val);
+        });
+
+        // finally update this.values
+        foreignKeys?.forEach(({ field }) => {
+          const val = fields[field] as unknown as string;
+          if (!val) {
+            return;
+          }
+          fkValues.set(field, val);
```

Comment:
This is inconsistent with the comment, and from what I can tell `fkValues` is not read again after writing it here.  Is this supposed to be updating something that persists across the `added()` `changed()` `removed()` calls?  maybe `this.values`?

Reply:
Oh, never mind -- `fkValues` is a reference to the value stored in `this.values` for `id`, so updating that Map is exactly what we want here.  I've been in static single assignment land for too long